public/app/features/explore/Logs/LogsContainer.tsx

```diff
@@ -59,45 +60,71 @@ interface LogsContainerProps extends PropsFromRedux {
 
 interface LogsContainerState {
   logDetailsFilterAvailable: boolean;
+  /**
+   * LogContext support for mixed data sources.
+   * @alpha
+   */
+  logContextSupport: Record<string, DataSourceApi<DataQuery> & DataSourceWithLogsContextSupport<DataQuery>>;
 }
 
 class LogsContainer extends PureComponent<LogsContainerProps, LogsContainerState> {
   state: LogsContainerState = {
     logDetailsFilterAvailable: false,
+    logContextSupport: {},
   };
 
   componentDidMount() {
-    this.checkFiltersAvailability();
+    this.checkDataSourcesFeatures();
   }
 
   componentDidUpdate(prevProps: LogsContainerProps) {
-    this.checkFiltersAvailability();
+    this.checkDataSourcesFeatures();
   }
 
-  private checkFiltersAvailability() {
+  private checkDataSourcesFeatures() {
     const { logsQueries, datasourceInstance } = this.props;
 
-    if (!logsQueries) {
+    if (!logsQueries || !datasourceInstance) {
       return;
     }
 
-    if (datasourceInstance?.modifyQuery || hasToggleableQueryFiltersSupport(datasourceInstance)) {
-      this.setState({ logDetailsFilterAvailable: true });
+    let newState: LogsContainerState = { ...this.state, logDetailsFilterAvailable: false };
+
+    // Not in mixed mode.
+    if (datasourceInstance.name !== MIXED_DATASOURCE_NAME) {
+      if (datasourceInstance?.modifyQuery || hasToggleableQueryFiltersSupport(datasourceInstance)) {
+        newState.logDetailsFilterAvailable = true;
+      }
+      if (hasLogsContextSupport(datasourceInstance)) {
+        logsQueries.forEach(({ refId }) => {
+          newState.logContextSupport[refId] = datasourceInstance;
+        });
+      }
+      this.setState(newState);
       return;
     }
 
+    // Mixed mode.
     const promises = [];
+    const refIds: string[] = [];
     for (const query of logsQueries) {
-      if (query.datasource) {
+      if (query.datasource && !newState.logContextSupport[query.refId]) {
         promises.push(getDataSourceSrv().get(query.datasource));
+        refIds.push(query.refId);
       }
     }
 
     Promise.all(promises).then((dataSources) => {
-      const logDetailsFilterAvailable = dataSources.some(
+      newState.logDetailsFilterAvailable = dataSources.some(
         (ds) => ds.modifyQuery || hasToggleableQueryFiltersSupport(ds)
       );
-      this.setState({ logDetailsFilterAvailable });
+      dataSources.forEach((ds, i) => {
+        if (hasLogsContextSupport(ds)) {
+          newState.logContextSupport[refIds[i]] = ds;
+        }
+      });
```

Comment:
Just a nit, but in the worst case this will loop through all `dataSources` twice. You could just add this to the `forEach` loop:

```
newState.logDetailsFilterAvailable = newState.logDetailsFilterAvailable || ds.modifyQuery || hasToggleableQueryFiltersSupport(ds);
```

Reply:
Updated.